internal/servicelog/write_buffer.go

```diff
@@ -0,0 +1,354 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package servicelog
+
+import (
+	"errors"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type StreamID int
+
+var (
+	Unknown StreamID = 0
+	Stdout  StreamID = 1
+	Stderr  StreamID = 2
+)
+
+func (s StreamID) String() string {
+	switch s {
+	case Stdout:
+		return "stdout"
+	case Stderr:
+		return "stderr"
+	default:
+		return "unknown"
+	}
+}
+
+type WriteIndex int32
+
+var (
+	TailIndex WriteIndex = -1
+)
+
+type write struct {
+	time       time.Time
+	start, end RingPos
+	streamID   StreamID
+	index      WriteIndex
+	ref        int32
+}
+
+func (l *write) empty() bool {
+	return l.time == time.Time{} && l.start == RingPos(0) && l.end == RingPos(0)
+}
+
+func (l *write) length() int {
+	return int(l.end - l.start)
+}
+
+func (l *write) acquire() int {
+	return int(atomic.AddInt32(&l.ref, 1))
+}
+
+func (l *write) release() int {
+	return int(atomic.AddInt32(&l.ref, -1))
+}
+
+func (l *write) refs() int {
+	return int(atomic.LoadInt32(&l.ref))
+}
+
+type WriteBuffer struct {
+	ringBuffer      *RingBuffer
+	writes          []write
+	placeholderTail write
+
+	writeMutex   sync.Mutex
+	writeClosed  bool
+	releaseMutex sync.RWMutex
+	signalMut    sync.RWMutex
+	signalChan   chan struct{}
+
+	tailIndex int32
+	headIndex int32
+	numWrites int32
+}
+
+var (
+	ErrNoMoreLines = errors.New("no more lines")
+)
+
+func NewWriteBuffer(maxWrites, maxBytes int) *WriteBuffer {
+	wb := WriteBuffer{
+		writes:     make([]write, maxWrites),
+		ringBuffer: NewRingBuffer(maxBytes),
+		tailIndex:  0,
+		headIndex:  -1,
+		numWrites:  0,
+		signalChan: make(chan struct{}),
+		placeholderTail: write{
+			index: TailIndex,
+		},
+	}
+	return &wb
+}
+
+func (wb *WriteBuffer) Close() error {
+	wb.writeMutex.Lock()
+	defer wb.writeMutex.Unlock()
+	if wb.writeClosed {
+		return nil
+	}
+	wb.writeClosed = true
+	wb.releaseWaiters()
+	return nil
+}
+
+func (wb *WriteBuffer) StreamWriter(streamID StreamID) io.Writer {
+	return &streamWriter{wb, streamID}
+}
+
+func (wb *WriteBuffer) Write(p []byte, streamID StreamID) (int, error) {
+	written := 0
+	defer func() {
+		if written > 0 {
+			wb.signalWaiters()
+		}
+	}()
+
+	writeTime := time.Now().UTC()
+
+	wb.writeMutex.Lock()
+	defer wb.writeMutex.Unlock()
+
+	if wb.writeClosed {
+		return 0, io.ErrClosedPipe
+	}
+
+	writeLength := len(p)
+	if writeLength > wb.ringBuffer.Capacity() {
+		writeLength = wb.ringBuffer.Capacity()
+	}
+
+	for writeLength > wb.ringBuffer.Free() || !wb.canAdvanceHead() {
+		err := wb.releaseTail()
+		if err != nil {
+			return written, err
+		}
+	}
+
+	index := WriteIndex(atomic.LoadInt32(&wb.headIndex) + 1)
+	write := wb.getWrite(index)
+	write.index = index
+	write.time = writeTime
+	write.streamID = streamID
+	write.start = wb.ringBuffer.Pos()
+	n, err := wb.ringBuffer.Write(p[:writeLength])
+	if err != nil && err != io.ErrShortWrite {
+		return 0, err
+	}
+	write.end = wb.ringBuffer.Pos()
+	written = n
```

Comment:
This file is where I got a bit lost. I don't understand what this piece does in the puzzle, since I was assuming all we needed was a ring buffer, and something similar to an iterator that consumes data from it. To have multiple consumers I imagine we need a list of earliest reads. So in comparison this all looks surprisingly complex.

There are even sleeps in the middle of that logic, which is a strong smell.

I'm hoping we can significantly simplify this.

Reply:
@niemeyer WriteBuffer tracks each write that comes out of stdout/stderr for a process, keeping track of the metadata around the write (ts, which stream, where it is in the ring buffer). It also handles multiple writer/multiple readers and the concurrency issues around that.

My original implementation was much more complicated as you saw. I've made an attempt to simplify it with a list of the iterators following the buffer, so that we can both see what are the earliest reads and keep a notification channel for each iterator.

Reply:
I see what it does. I don't understand why it does that, though. Why isn't this just the ring buffer itself with just tracking of read position?  We do need to track the position of readers, but that's just a RingBufferReader with an int (or RingPos) of its current position, that once Closed gets off the list.

For timestamps, we can add them when we write the data into the buffer, as logging normally does.

For finding out which stream, I don't think we want that per earlier comment. Separating stdout and stderr creates bad logging because they get out of sync. To fix that we need to indeed pass a single file descriptor into the process (in Go, cmd.Stdout and cmd.Stderr need to be the same buffer). So tracking the stream ID is impossible if we don't even have two streams.

Tracking individual writes: why?   I could understand tracking individual lines and having all of our API be line-based, but we are neither benefiting from a simple API that manipulates a list of strings, nor a simple API that manipulates a buffer.

Happy to have a conversation about this with more bandwidth.

Reply:
FWIW, I believe in many processes stdout is buffered and stderr is not, which means lines get out of sync anyway.
I would certainly do a best effort to try and keep them in sync (don't make the problem worse), but also recognize that they are fundamentally 2 streams that do get out of sync.

Reply:
I don't understand what that means, given that when you create a process you're the one defining what the stdout and stderr file descriptors are. If it is the same buffer, they cannot get out of sync because there is only one stream and one buffer. If they are out of sync, the only way to get that is if some child processes spawned further processes and messed that up, in which case Pebble has no saying and it's not its problem anymore.

Reply:
@jameinel Your email reply got off thread in GitHub, so I'm bringing it back:

> So inside the child process, they may wrap the raw file handle in a buffered writer or not. I have definitely seen applications which do buffer their writes to stdout but not to stderr. I do believe there isn't much reason for Pebble to pass in 2 different file handles if we are going to be putting the content into a single stream/buffer anyway.

This is the case I described above: if an application misbehaves internally, of course there's nothing we can do. The discussion here is about getting our part right.

Reply:
### Single stream
Interesting. So by having a single fd we can guarantee write order on both stdout/stderr because they **are** the same. In doing so we lose the ability to distinguish between stdout/stderr. I don't think that is too much of an issue, it does limit what level of filtering the end user can do, but I don't think that is too much of a concern. `kubectl logs` for example doesn't let you separate the two, they use a single file for both stdout/stderr.

Separate FDs does make more sense in a `ssh`/`exec` scenario, when we add the ability to run commands via pebble, but in that case we don't care about timestamps, we can just use two ring buffers.

### Embedding timestamps
If we embed the timestamps in the ring byte buffer, it will make multiple log formats a little bit more difficult. The log format I personally care about the most is a **raw** output, so I can see the exact output of the program (sans timestamps/service name), so if the encoding in the buffer is strict (as to be able to recreate the output as it was received), then I don't see any problem with this.

There is a tiny amount of state keeping for insertion of timestamp, but it can be as simple as `on the next byte after \n insert the timestamp`. The state is required because a line could span multiple write on the fd.

Streamed output from a program with long lines (spanned over multiple `write`s) makes it difficult to represent the API in the form of lines, I think a streaming style API will better suit this. Within go this would be an io.Reader and over HTTP its just as simple. A line by line API would have to omit incomplete lines when streaming/following, this is probably fine.

With embedded timestamps, mixing multiple streams (multiple services' output) will be a little bit more difficult, will need to read ahead on all the io.Reader for each service, then select the most recent by parsing the timestamps and comparing.
I think we can use the same iterator interface we currently have, the iterator can handle all this.

Reply:
I'm on board with having a single stream - I don't think that's too useful. However, I do prefer the timestamps being "metadata" rather than string data in the ring buffer that clients may have to parse out if they want to reformat. This is a tricky discussion to have async. I suggest we discuss and finalize this in a video chat -- I'll send a calendar invite / Mattermost message to see if we can touch base today.